source/extensions/filters/http/cache/hazelcast_http_cache/hazelcast_http_cache.cc

```diff
@@ -0,0 +1,178 @@
+  ENVOY_LOG(info,
+            "Cache statistics can be observed on Hazelcast Management Center"
+            " from the map named {}.",
+            unified_ ? cluster_accessor.responseMapName() : cluster_accessor.headerMapName());
+}
+
+void HazelcastHttpCache::shutdown(bool destroy) {
+  if (!accessor_) {
+    ENVOY_LOG(warn, "Cache is already offline.");
+    return;
+  }
+  if (accessor_->isRunning()) {
+    ENVOY_LOG(info, "Shutting down Hazelcast connection...");
+    accessor_->disconnect();
+    ENVOY_LOG(info, "Cache is offline now.");
+  } else {
+    ENVOY_LOG(warn, "Hazelcast client is already disconnected.");
+  }
+  if (destroy) {
+    accessor_.reset();
+  }
+}
+
+HazelcastHttpCache::~HazelcastHttpCache() { shutdown(true); }
+
+LookupContextPtr HazelcastHttpCache::makeLookupContext(LookupRequest&& request) {
+  if (unified_) {
+    return std::make_unique<UnifiedLookupContext>(*this, std::move(request));
+  } else {
+    return std::make_unique<DividedLookupContext>(*this, std::move(request));
+  }
+}
+
+InsertContextPtr HazelcastHttpCache::makeInsertContext(LookupContextPtr&& lookup_context) {
+  ASSERT(lookup_context != nullptr);
+  if (unified_) {
+    return std::make_unique<UnifiedInsertContext>(*lookup_context, *this);
+  } else {
+    return std::make_unique<DividedInsertContext>(*lookup_context, *this);
+  }
+}
+
+// TODO(enozcan): Implement when it's ready on the filter side.
+//  Depending on the filter's implementation, the cached entry's
+//  variant_key_ must be updated as well. Also, if vary headers
+//  change then the hash key of the response will change and
+//  updating only header map will not be enough in this case.
+void HazelcastHttpCache::updateHeaders(LookupContextPtr&&, Http::ResponseHeaderMapPtr&&) {
+  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+}
+
+constexpr absl::string_view HazelcastCacheName = "envoy.extensions.http.cache.hazelcast";
+
+// Cluster wide cache statistics should be observed on Hazelcast Management Center.
+// They are not stored locally.
+CacheInfo HazelcastHttpCache::cacheInfo() const {
+  CacheInfo cache_info;
+  cache_info.name_ = HazelcastCacheName;
+  cache_info.supports_range_requests_ = true;
+  return cache_info;
+}
+
+std::string HazelcastHttpCacheFactory::name() const { return std::string(HazelcastCacheName); }
+
+ProtobufTypes::MessagePtr HazelcastHttpCacheFactory::createEmptyConfigProto() {
+  return std::make_unique<HazelcastHttpCacheConfig>();
+}
+
+HttpCache& HazelcastHttpCacheFactory::getCache(
+    const envoy::extensions::filters::http::cache::v3alpha::CacheConfig& config) {
+  if (!cache_) {
+    HazelcastHttpCacheConfig hz_cache_config;
+    MessageUtil::unpackTo(config.typed_config(), hz_cache_config);
+    cache_ = std::make_unique<HazelcastHttpCache>(hz_cache_config);
+    cache_->start();
+  }
+  return *cache_;
+}
+
+HazelcastHttpCache* HazelcastHttpCacheFactory::getOfflineCache(
```

Comment:
This should return HazelcastHttpCachePtr.

Reply:
Done